application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java

```diff
@@ -139,26 +149,78 @@ private boolean handleChecks(@NotNull Member bot, @NotNull Member author,
+        List<List<ActionRecord>> groupedActions = new ArrayList<>();
+        for (int i = 0; i < actions.size(); i++) {
+            if (i % AuditCommand.MAX_PAGE_LENGTH == 0) {
+                groupedActions.add(new ArrayList<>(AuditCommand.MAX_PAGE_LENGTH));
+            }
 
-        MessageEmbed summary = createSummaryMessage(user, actions);
-        if (actions.isEmpty()) {
-            event.replyEmbeds(summary).queue();
-            return;
+            groupedActions.get(groupedActions.size() - 1).add(actions.get(i));
         }
 
-        // Computing messages for actual actions is done deferred and might require asking the
-        // Discord API
-        event.deferReply().queue();
-        JDA jda = event.getJDA();
+        return groupedActions;
+    }
 
-        RestAction<List<MessageEmbed>> messagesTask = RestAction
-            .allOf(actions.stream().map(action -> actionToMessage(action, jda)).toList());
-        messagesTask.map(messages -> prependElement(summary, messages))
-            .flatMap(messages -> event.getHook().sendMessageEmbeds(messages))
-            .queue();
+    private @NotNull Message auditUser(long guildId, long targetId, long callerId, int pageNumber,
+            @NotNull JDA jda) {
+        List<ActionRecord> actions = actionsStore.getActionsByTargetAscending(guildId, targetId);
+        List<List<ActionRecord>> groupedActions = groupActions(actions);
+        int totalPages = groupedActions.size();
+
+        EmbedBuilder audit = createSummaryEmbed(jda.retrieveUserById(targetId).complete(), actions);
+
+        if (groupedActions.isEmpty()) {
+            return new MessageBuilder(audit.build()).build();
+        }
+
+        groupedActions.get(pageNumber - 1)
+            .forEach(action -> audit.addField(actionToField(action, jda)));
+
+        return new MessageBuilder(audit.setFooter("Page: " + pageNumber + "/" + totalPages).build())
+            .setActionRows(makeActionRow(guildId, targetId, callerId, pageNumber, totalPages))
+            .build();
+    }
+
+    private @NotNull ActionRow makeActionRow(long guildId, long targetId, long callerId,
+            int pageNumber, int totalPages) {
+        String stringGuildId = String.valueOf(guildId);
+        String stringTargetId = String.valueOf(targetId);
+        String stringCallerId = String.valueOf(callerId);
+        String stringPageNumber = String.valueOf(pageNumber);
+
+        Button previousButton = Button.primary(generateComponentId(stringGuildId, stringTargetId,
+                stringCallerId, stringPageNumber, "-1"), "⬅");
+        if (pageNumber == 1) {
+            previousButton = previousButton.asDisabled();
+        }
+
+        Button nextButton = Button.primary(generateComponentId(stringGuildId, stringTargetId,
+                stringCallerId, stringPageNumber, "1"), "➡");
+        if (pageNumber == totalPages) {
+            nextButton = nextButton.asDisabled();
+        }
```

Comment:
I do not like how this code section is still almost 100% identical. This is code duplication. Introduce a helper method that creates a button with a label and a turnPagesBy argument and then use that:

```java
private static Button createPageTurnButton(String label, int turnPagesBy, ...) {
  return Button.primary(...)
}
```

Reply:
it'd look the same, even if i introduced that new method

Reply:
But you get rid of the duplication. Lets say you want to make both buttons red, currently you need to adjust 2 locations and worst case you forget one and they are out of sync...

By introducing a helper, you merge both into a single code location and reduce the duplication. That will also make Sonar happy, which hates code duplication.